### PR TITLE
Receivable SDK V2

### PR DESCRIPTION
**Pull request by mliu**

### CODE REVIEW

File: packages/examples/src/createReceivableAlternateNetworkV2.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  const TEST_PRIVATE_KEY = process.env.TEST_PRIVATE_KEY
+  const provider = new ethers.providers.JsonRpcProvider(
+    `http://localhost:8545`,
+    {
+      name: 'Localhost',
+      chainId: ChainEnum.Localhost,
+    },
+  )
+  const wallet = new Wallet(TEST_PRIVATE_KEY, provider)
+  const adminWallet = new Wallet(
```
Comment (shan-57blocks) [inline packages/examples/src/createReceivableAlternateNetworkV2.ts:16]: adminWallet is not used

File: packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import { getContract } from '../utils'
+
+/**
+ * Returns an ethers contract instance for the V2 Receivable contract
+ * associated with the given pool name on the current chain.
+ *
+ * @param {ethers.providers.Provider | ethers.Signer} signerOrProvider The provider or signer instance to use for the contract.
+ * @param {number} chainId The chain id where the contract instance exists
+ * @returns {Contract | null} A contract instance for the RealWorldReceivable contract or null if it could not be found.
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts:17]: Should we still name it 'RealWorldReceivable' or just 'Receivable'?

File: packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  const poolConfigContract = await getPoolConfigContractV2(poolName, provider)
+
+  if (!poolConfigContract) {
+    throw new Error('Could not find PoolConfig contract')
+  }
+
+  const receivableAsset = await poolConfigContract.receivableAsset()
+
+  return getContract(receivableAsset, RECEIVABLE_V2_ABI, signerOrProvider)
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts:38]: Can cast the Contract type to Receivable Contract type

File: packages/huma-sdk/src/services/v2/HumaContext.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import { JsonRpcProvider, Web3Provider } from '@ethersproject/providers'
+import { POOL_NAME, POOL_TYPE } from '@huma-finance/shared'
+import { ethers } from 'ethers'
+
+export class HumaContext {
+  #signer: ethers.Signer
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaContext.ts:6]: From what I observed, normally it's named as _signer.

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+  constructor({
+    humaContext,
+    arWeavePaymentChainId,
+  }: {
+    humaContext: HumaContext
+    arWeavePaymentChainId?: ChainEnum
+  }) {
+    if (!humaContext) {
+      throw new Error('All parameters are required')
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:26]: The error message is a little bit confusing.

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    this.#humaContext = humaContext
+    this.#arWeavePaymentChainId = arWeavePaymentChainId
+  }
+
+  async createReceivableWithMetadata(
+    privateKey: string,
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:34]: Can we name it like  aRWeavePrivateKey so not messing with wallet private key?

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    referenceId: string = '',
+    extraTags: { name: string; value: string }[] = [],
+    lazyFund: boolean = true,
+    gasOpts?: Overrides,
+  ): Promise<TransactionResponse> {
+    if (typeof metadata !== 'object' || Array.isArray(metadata)) {
+      throw new Error('Input must be a JSON object.')
+    }
+
+    this.throwIfReferenceIdExists(referenceId)
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:48]: Add await?

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+  async createReceivable(
+    currencyCode: number,
+    receivableAmount: BigNumber,
+    maturityDate: number,
+    metadataURI: string,
+    referenceId: string = '',
+    gasOpts: Overrides = {},
+  ): Promise<TransactionResponse> {
+    this.throwIfReferenceIdExists(referenceId)
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:76]: Add await

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+      { name: 'appName', value: 'HumaFinance' },
+      { name: 'ownerAddress', value: signerAddress },
+      { name: 'poolName', value: this.#humaContext.poolName },
+      { name: 'poolType', value: this.#humaContext.poolType },
+      { name: 'referenceId', value: referenceId },
+      ...extraTags,
+    ]
+
+    // Append referenceId to metadata (if it's not already there)
+    if (!Object.prototype.hasOwnProperty.call(metadata, 'referenceId')) {
```
Comment (shan-57blocks) [inline packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:136]: I think metadata.hasOwnProperty('referenceId') is simpler.

**Review by shan-57blocks — APPROVED** [review]

File: packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts (hunk 2 above)
Comment (mliu) [reply packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts:17]: Accidentally copied the wrong comment, updating

File: packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts (hunk 3 above)
Comment (mliu) [reply packages/huma-sdk/src/helpers/ReceivableV2ContractHelper.ts:38]: Will do this in the followup PR.

File: packages/huma-sdk/src/services/v2/HumaContext.ts (hunk 4 above)
Comment (mliu) [reply packages/huma-sdk/src/services/v2/HumaContext.ts:6]: _signer actually violates the ESLint rule no-underscore-dangle. I was reading the ESLint documentation for this (https://eslint.org/docs/latest/rules/no-underscore-dangle) and it was recommended to use the private properties standard introduced in EMCAScript 2022. I think it's fine to adopt this unless you have any other thoughts.

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts (hunk 6 above)
Comment (mliu) [reply packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:34]: This actually is a wallet private key, it's used to pay for arweave uploads. But I will rename it to be more clear

File: packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts (hunk 7 above)
Comment (mliu) [reply packages/huma-sdk/src/services/v2/HumaReceivableFactory.ts:48]: Nice catch

**Pull request merged**